src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+/********************************************************************++
+Copyright (c) Microsoft Corporation.  All rights reserved.
+--********************************************************************/
+
+using System.Linq;
+using System.IO;
+using System.Text;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Management.Automation.Internal;
+using System.Management.Automation.Language;
+using Microsoft.PowerShell.DesiredStateConfiguration.Internal;
+using System.Reflection;
+using System.Linq.Expressions;
+
+// Keyword and its attributes are in the Automation namespace to make them consistent
+// with Cmdlet and its attributes; although Keyword does not currently inherit from
+// InternalCommand, it conceivably will when appropriate
+
+namespace System.Management.Automation
+{
+    #region Keyword Command Class
+
+    /// <summary>
+    /// An abstract class to extend in order to write a keyword specification. Give this class the KeywordAttribute
+    /// in order to use it as a PowerShell DynamicKeyword specification. Any properties with
+    /// the KeywordParameterAttribute will be parameters, while properties with the KeywordPropertyAttribute will
+    /// be keyword properties.
+    /// </summary>
+    public abstract class Keyword
+    {
+        /// <summary>
+        /// Constructs a keyword with null runtime delegates. This is intended to be overridden.
+        /// </summary>
+        protected Keyword()
+        {
+        }
+
+        /// <summary>
+        /// Specifies the call to run on a DynamicKeyword data object before the AST node
+        /// containing that keyword is parsed
+        /// </summary>
+        public Func<DynamicKeyword, ParseError[]> PreParse { get; set; }
+
+        /// <summary>
+        /// Specifies the call to run on a DynamicKeyword statement AST immediately after that AST node
+        /// has been parsed
+        /// </summary>
+        public Func<DynamicKeywordStatementAst, ParseError[]> PostParse { get; set; }
+
+        /// <summary>
+        /// Specifies the call to run on a DynamicKeyword statement AST node at semantic check time (to perform
+        /// any user-specified semantic checks)
+        /// </summary>
+        public Func<DynamicKeywordStatementAst, ParseError[]> SemanticCheck { get; set; }
+    }
+
+    #endregion /* Keyword Command Class */
+
+    #region Dynamic Keyword Specification Attributes
+
+    /// <summary>
+    /// Specifies that a class describes a DynamicKeyword for PowerShell
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class)]
+    public class KeywordAttribute : ParsingBaseAttribute
+    {
+        // Default attribute parameter values
+
+        /// <summary>
+        /// Default keyword name mode
+        /// </summary>
+        public const DynamicKeywordNameMode DefaultNameMode = DynamicKeywordNameMode.NoName;
```

Comment:
Agreed. The value needs to be shared by the metadata reader in some way though, so we can either make them internal and share the values, or instantiate the attribute when it's read to get the defaults. The second might be cleaner, but less efficient.